### PR TITLE
add reconnect test

### PR DESCRIPTION
**Pull request by StalderT**

### CODE REVIEW

**Comment by StalderT** [comment]
fails as expected :
```
22: [01/01/1601 00:20:01.001] error/client	GetEndpointRequest failed with error code BadSecureChannelIdInvalid
22: [01/01/1601 00:20:01.001] info/channel	Connection 10 | SecureChannel 2 | CloseSecureChannel
22: [01/01/1601 00:20:01.001] info/network	Shutting down the TCP network layer
22: [01/01/1601 00:20:01.001] info/network	Connection 10 | Closed by the server
22: 66%: Checks: 3, Failures: 1, Errors: 0
22: /home/travis/build/open62541/open62541/tests/client/check_client_securechannel.c:124:F:Client SecureChannel:SecureChannel_reconnect:0: Assertion 'retval==0x00' failed: retval==2149711872, 0x00==0
22/27 Test #22: check_client_securechannel ............***Failed    0.07 sec
```

**Comment by jpfr** [comment]
The new test manually sets the state, an internal value of the client.
The state should be (and is) UA_CLIENTSTATE_SESSION.
I will try to replicate the issue by only using "external" interactions with the client.

**Comment by StalderT** [comment]
I agree. Just to explain, in certain case (network failure), the client call UA_Client_disconnect. UA_Client_disconnect try to close the session but due to network error never arrive to the server and still opened in server. When the client try to reconnect an error occur. Forcing client->state = UA_CLIENTSTATE_CONNECTED; simulate this case.

**Comment by Pro** [comment]
@jpfr you probably want to rebase on master. There was another commit after your rebase which ignores the failing securechannel unit test

**Comment by jpfr** [comment]
Found and fixed.
When opening a new SecureChannel we ran "GetEndpoints" with the old session authentication token.
Now, when we have an old Session, we try to recover it and do not try GetEndpoints at all.

Still, there are probably more edge cases not covered in the client.
More unit tests are always good here. 😄 

@StalderT Okay. I understand now.
I added a comment to explain that we force-close the SecureChannel.

@Pro Yes. Thanks. Rebased. I added a "sleep" that will hopefully make the unit test run stable.
Let's see whether travis is stable.

**Comment by StalderT** [comment]
Ok but If the server restart, then the client try to reconnect. The activate session will fail because she doesn't exists (same if session close the session timout by example).

**Comment by Pro** [comment]
@jpfr I rebased again on current master to reenable the timeout unit test.
Anyways I don't think that the sleep fixed the issue because I had the problem of #1388 even after that fix #1385 
Let's see if Travis is happy for this PR, and if yes let's restart travis for a few times to see if it is really fixed.

**Comment by StalderT** [comment]
Something like that ?

```
UA_StatusCode
UA_Client_connectInternal(UA_Client *client, const char *endpointUrl,
                          UA_Boolean endpointsHandshake, UA_Boolean createNewSession) {
    UA_ChannelSecurityToken_init(&client->channel.securityToken);
    client->channel.state = UA_SECURECHANNELSTATE_FRESH;

    if(client->state >= UA_CLIENTSTATE_CONNECTED)
        return UA_STATUSCODE_GOOD;

    UA_StatusCode retval = UA_STATUSCODE_GOOD;
    client->connection =
        client->config.connectionFunc(client->config.localConnectionConfig,
                                      endpointUrl, client->config.timeout);
    if(client->connection.state != UA_CONNECTION_OPENING) {
        retval = UA_STATUSCODE_BADCONNECTIONCLOSED;
        goto cleanup;
    }

    UA_String_deleteMembers(&client->endpointUrl);
    client->endpointUrl = UA_STRING_ALLOC(endpointUrl);
    if(!client->endpointUrl.data) {
        retval = UA_STATUSCODE_BADOUTOFMEMORY;
        goto cleanup;
    }

    /* Open a TCP connection */
    client->connection.localConf = client->config.localConnectionConfig;
    retval = HelAckHandshake(client);
    if(retval != UA_STATUSCODE_GOOD)
        goto cleanup;
    client->state = UA_CLIENTSTATE_CONNECTED;

    /* Open a SecureChannel. TODO: Select with endpoint  */
    client->channel.connection = &client->connection;
    retval = openSecureChannel(client, false);
    if(retval != UA_STATUSCODE_GOOD)
        goto cleanup;
    client->state = UA_CLIENTSTATE_SECURECHANNEL;

    if(createNewSession && (!UA_NodeId_equal(&client->authenticationToken, &UA_NODEID_NULL))) {
        /* Try to activate the Session for this SecureChannel */
        retval = activateSession(client);
        if(retval == UA_STATUSCODE_GOOD){
            client->state = UA_CLIENTSTATE_SESSION;
            return retval;
        }
        /* Can't activate session using actual authenticationToken
         * we remove it */
        UA_NodeId_deleteMembers(&client->authenticationToken);
        UA_NodeId_init(&client->authenticationToken);        
    }

    /* Get Endpoints */
    if(endpointsHandshake) {
        retval = getEndpoints(client);
        if(retval != UA_STATUSCODE_GOOD)
            goto cleanup;
    }

    /* Open a Session */
    if(createNewSession) {
        retval = createSession(client);
        if(retval != UA_STATUSCODE_GOOD)
            goto cleanup;
        retval = activateSession(client);
        if(retval != UA_STATUSCODE_GOOD)
            goto cleanup;
        client->state = UA_CLIENTSTATE_SESSION;
        return UA_STATUSCODE_GOODPOSTACTIONFAILED; // see the explanations that follow
    }
    return retval;

cleanup:
    UA_Client_disconnect(client);
    return retval;
}
```

If the client application use subscription, UA_Client_connect sould indicate if a new session was create : 

```
    while (running) {
        /* if already connected, this will return GOOD and do nothing */
        /* if the connection is closed/errored, the connection will be reset and then reconnected */
        /* Alternatively you can also use UA_Client_getState to get the current state */
        /* if a new session was create, this will return GOODPOSTACTIONFAILED */
        UA_StatusCode retval = UA_Client_connect(client, "opc.tcp://localhost:4840");

        if ((retval != UA_STATUSCODE_GOOD) && (retval != UA_STATUSCODE_GOODPOSTACTIONFAILED)) {
            UA_LOG_ERROR(logger, UA_LOGCATEGORY_CLIENT, "Not connected. Retrying to connect in 1 second");
            /* The connect may timeout after 1 second (see above) or it may fail immediately on network errors */
            /* E.g. name resolution errors or unreachable network. Thus there should be a small sleep here */
            UA_sleep_ms(1000);
            continue;
        }

        if (retval == UA_STATUSCODE_GOODPOSTACTIONFAILED) {
            /* A new session was create, create/recreate subscription */

            /* Create a subscription */
            UA_UInt32 subId = 0;
            UA_Client_Subscriptions_new(client, UA_SubscriptionSettings_default, &subId);
            if(subId)
                printf("Create subscription succeeded, id %u\n", subId);
            /* Add a MonitoredItem */
            UA_NodeId monitorThis = UA_NODEID_STRING(1, "the.answer");
            UA_UInt32 monId = 0;
            UA_Client_Subscriptions_addMonitoredItem(client, subId, monitorThis, UA_ATTRIBUTEID_VALUE,
                                                     &handler_TheAnswerChanged, NULL, &monId, 250);
            if (monId)
                printf("Monitoring 'the.answer', id %u\n", subId);
        }

        UA_Client_Subscriptions_manuallySendPublishRequest(client);

         if ((retval != UA_STATUSCODE_GOOD) && (retval != UA_STATUSCODE_GOODNONCRITICALTIMEOUT)) {
            UA_LOG_ERROR(logger, UA_LOGCATEGORY_CLIENT, "Connection was closed. Reconnecting ...");
            continue;
        }
        UA_sleep_ms(1000);
    }
```

Can you confirm, if the session is reactivated the subscription are not lost ?

We can use another status code than UA_STATUSCODE_GOODPOSTACTIONFAILED

**Comment by StalderT** [comment]
I have an idea about reconnection. Tomorrow, I'll make a proposal.

File: tests/client/check_client_securechannel.c [hunk 1]
```diff
@@ -100,23 +100,6 @@ START_TEST(SecureChannel_reconnect) {
 }
 END_TEST
 
-START_TEST(SecureChannel_reconnect) {
```
Comment (jpfr) [inline tests/client/check_client_securechannel.c]: You cannot remove the test and still reference it below.

Comment (jpfr) [reply tests/client/check_client_securechannel.c]: Sorry, didn't see the duplicate warning.

**Comment by StalderT** [comment]
I added callbacks. I wanted to have your opinion before continuing in this way.

**Comment by jpfr** [comment]
Hey @StalderT ,

I like the initiative. Still, we are mixing concerns in one PR.

Can we please merge #1397 first? It contains just the first commits from this PR.
They do **the right thing** even if we have a timing condition in the unit tests.

Then, we can rebase this PR and call it "client connect callback" or similar.
Can you explain what are the differences to #1117? The goals seem pretty similar.

**Comment by StalderT** [comment]
ok first fix/merge https://github.com/open62541/open62541/pull/1392 and https://github.com/open62541/open62541/pull/1397 then close #1386

For https://github.com/open62541/open62541/pull/1397 we need to use 
```
UA_StatusCode
UA_Client_connectInternal(UA_Client *client, const char *endpointUrl,
                          UA_Boolean endpointsHandshake, UA_Boolean createNewSession) {
    if(client->state >= UA_CLIENTSTATE_CONNECTED)
        return UA_STATUSCODE_GOOD;

    UA_ChannelSecurityToken_init(&client->channel.securityToken);
    client->channel.state = UA_SECURECHANNELSTATE_FRESH;

    UA_StatusCode retval = UA_STATUSCODE_GOOD;
    client->connection =
        client->config.connectionFunc(client->config.localConnectionConfig,
                                      endpointUrl, client->config.timeout);
    if(client->connection.state != UA_CONNECTION_OPENING) {
        retval = UA_STATUSCODE_BADCONNECTIONCLOSED;
        goto cleanup;
    }

    UA_String_deleteMembers(&client->endpointUrl);
    client->endpointUrl = UA_STRING_ALLOC(endpointUrl);
    if(!client->endpointUrl.data) {
        retval = UA_STATUSCODE_BADOUTOFMEMORY;
        goto cleanup;
    }

    /* Open a TCP connection */
    client->connection.localConf = client->config.localConnectionConfig;
    retval = HelAckHandshake(client);
    if(retval != UA_STATUSCODE_GOOD)
        goto cleanup;
    client->state = UA_CLIENTSTATE_CONNECTED;

    /* Open a SecureChannel. TODO: Select with endpoint  */
    client->channel.connection = &client->connection;
    retval = openSecureChannel(client, false);
    if(retval != UA_STATUSCODE_GOOD)
        goto cleanup;
    client->state = UA_CLIENTSTATE_SECURECHANNEL;

    if(createNewSession && (!UA_NodeId_equal(&client->authenticationToken, &UA_NODEID_NULL))) {
        /* Try to activate the Session for this SecureChannel */
        retval = activateSession(client);
        if(retval == UA_STATUSCODE_GOOD){
            client->state = UA_CLIENTSTATE_SESSION;
            return retval;
        }
    }

    UA_NodeId_deleteMembers(&client->authenticationToken);
    UA_NodeId_init(&client->authenticationToken);

    /* Get Endpoints */
    if(endpointsHandshake) {
        retval = getEndpoints(client);
        if(retval != UA_STATUSCODE_GOOD)
            goto cleanup;
    }

    /* Open a Session */
    if(createNewSession) {
        retval = createSession(client);
        if(retval != UA_STATUSCODE_GOOD)
            goto cleanup;
        retval = activateSession(client);
        if(retval != UA_STATUSCODE_GOOD)
            goto cleanup;
        client->state = UA_CLIENTSTATE_SESSION;
    }
    return retval;

cleanup:
    UA_Client_disconnect(client);
    return retval;
}
```

**Comment by StalderT** [comment]
For info https://github.com/open62541/open62541/commit/23cc7ff34d352934d60aca8492ec42cb218d3afa seem to fix timing issue in tests

**Comment by StalderT** [comment]
I watched https://github.com/open62541/open62541/pull/1117 and the goal seems to be the same.  Maybe the title is not very suitable.
